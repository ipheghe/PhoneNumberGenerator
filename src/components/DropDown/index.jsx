// react lib imports
import React from 'react';

// third-party-imports
import PropTypes from 'prop-types';

// styles
import './DropDown.scss';

/**
 * The renders the Drop Down component
 *
 * @returns {JSX} JSX
 */
const DropDown = ({ handleSortClick, className }) =>  (
  <div className={`dropdown__container ${className}`}>
    <button
      type="button"
      className="dropdown__btn"
    >
      <p>Sort By <span><i className="down__arrow"></i></span></p>
    </button>
    <div className="dropdown__content">
      <button
        type="button"
        className="dropdown__btn__item"
        onClick={(event) => handleSortClick(event, 'unique')}
      >
        Id
      </button>
      <button
        type="button"
        className="dropdown__btn__item"
        onClick={(event) => handleSortClick(event, 'highest')}
      >
        Highest Number
      </button>
      <button
        type="button"
        className="dropdown__btn__item"
        onClick={(event) => handleSortClick(event, 'lowest')}
      >
        Lowest Number
      </button>
    </div>
  </div>
);

DropDown.propTypes = {
  className: PropTypes.string,
  handleSortClick: PropTypes.func.isRequired,
};

DropDown.defaultProps = {
  className: '',
};

export default DropDown;
