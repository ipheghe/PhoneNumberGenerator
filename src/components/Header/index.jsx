import React from 'react';
import { Link } from '@reach/router';

import './Header.scss';

/**
 * The Header component primarily used for navigation between pages
 *
 * @returns {JSX} JSX
 */
const Header = () =>  (
  <div className="header__container">
    <h3>Number Generator</h3>
    <ul className="menu__container">
      <li className="menu__container__items">
        <Link to="/">Home</Link>
      </li>
      <li className="menu__container__items">
        <Link to="/operations">Operations</Link>
      </li>
    </ul>
  </div>
);

export default Header;
