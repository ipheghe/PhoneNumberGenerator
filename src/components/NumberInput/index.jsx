// react lib imports
import React from 'react';

// third-party-imports
import PropTypes from 'prop-types';

// styles
import './NumberInput.scss';

/**
 * The NumberInput component that renders an input and button element
 *
 * @returns {JSX} JSX
 */
const NumberInput = ({
  value,
  onChange,
  buttonText,
  handleButtonClick,
}) =>  (
  <div className="input__container">
    <input
      id="chosenNumber"
      className="input__field"
      type="number"
      name="chosenNumber"
      value={value}
      onChange={onChange}
      max="500"
      min="1"
    />
    <button
      className="input__button"
      type="button"
      onClick={handleButtonClick}
    >
      {buttonText}
    </button>
  </div>
);

NumberInput.propTypes = {
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]).isRequired,
  onChange: PropTypes.func.isRequired,
  buttonText: PropTypes.string.isRequired,
  handleButtonClick: PropTypes.func.isRequired,
};

export default NumberInput;
