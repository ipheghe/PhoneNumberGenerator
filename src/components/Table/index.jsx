// react Lib imports
import React from 'react';

// third-party imoports
import PropTypes from 'prop-types';

// styles
import './Table.scss';

/**
 * This component displays a table with props data and configurations
 *
 * @class Table
 */
const Table = ({ config, headers, data }) => {
  /**
   *  Renders Table headers
   *
   * @returns {JSX} JSX
   */
  const renderTableHeader = () => {
    return (
      <div className="table-row table-header">
        {headers.map(header => (
          <span
            key={header}
            className={`${config[header].width || 'table-col'}`}
          >
            {config[header].alias ? config[header].alias : header}
          </span>
        ))}
      </div>
    );
  };

  /**
   * Renders Table Body
   *
   * @returns {JSX} JSX
   */
  const renderTableBody = () => {
    const updatedData = data.map((item, index) => {
      item['serial'] = index + 1;
      return item;
    });

    return (
      <div>
        {updatedData && updatedData.map((row, id) => (
          <div className="table-row" key={id}>
            {headers.map(header => (
              <span
                className={`${config[header].width || 'table-col'}`}
                key={header}
              >
                {row[header]}
              </span>
            ))}
          </div>
        ))}
      </div>
    );
  };

  return (
    <div className="table-wrapper">
      {renderTableHeader()}
      {renderTableBody()}
    </div>
  );
};

Table.propTypes = {
  config: PropTypes.object,
  headers: PropTypes.array,
  data: PropTypes.array,
};

export default Table;
