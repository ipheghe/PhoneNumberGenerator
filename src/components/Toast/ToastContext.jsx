import React from 'react';

const ToastContext = React.createContext({});

export const Provider = ToastContext.Provider;
export const Consumer = ToastContext.Consumer;
