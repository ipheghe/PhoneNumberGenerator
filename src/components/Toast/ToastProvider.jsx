// react lib imports
import React, { Component } from 'react';

// styles
import './Toast.scss';

// components
import { Provider } from './ToastContext';

/**
 * The represents the Toast Provider component
 *
 * @returns {JSX} JSX
 */
export default class ToastProvider extends Component {
  state = {
    message: null,
    className: '',
  };

  /**
   * This method shows toast message
   *
   * @param {string} message - message
   * @param {string} className - className
   * @return {void}  null
   */
  showMessage = (message, className) => {
    this.setState({ message, className }, () => this.hideMessage());
  };

  /**
   * This method hides toast message
   *
   * @return {void}  null
   */
  hideMessage = () => {
    const { timeOut } = this.props;
    setTimeout(() => {
      this.setState({ message: null, className: '' });
    }, timeOut || 3000);
  };

  /**
   * Renders Toast Provider
   *
   * @returns {JSX} jsx
   */
  render() {
    const { children } = this.props;
    const { message, className } = this.state;
    return (
      <Provider value={{
        className,
        message,
        showMessage: this.showMessage,
        hideMessage: this.hidemessage,
      }}>
        {children}
      </Provider>
    );
  }
}
