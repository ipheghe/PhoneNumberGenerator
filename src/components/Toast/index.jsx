// react lib imports
import React from 'react';

// styles
import './Toast.scss';

// components
import { Consumer } from './ToastContext';

/**
 * This renders the Toast component
 *
 * @returns {JSX} JSX
 */
export const Toast = ({ message, className }) => (
  <>
    {message && (
      <div className={`toast__container ${className}`}>
        <p>{message}</p>
      </div>
    )}
  </>
);

/**
 * This renders the Toast component
 *
 * @param {object} props - props
 * @returns {JSX} JSX
 */
export default function ToastWithContext(props) {
  return (
    <Consumer>
      {context => (
        <Toast
          {...props}
          message={context.message}
          className={context.className}
        />
      )}
    </Consumer>
  );
}
