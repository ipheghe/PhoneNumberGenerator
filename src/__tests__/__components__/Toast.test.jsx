// react libraries
import React from 'react';

// third-party libraries
import { shallow } from 'enzyme';
import render from 'react-test-renderer';
import { spy } from 'sinon';

// components
import { Toast } from '../../components/Toast';
import ToastProvider from '../../components/Toast/ToastProvider';

const props = {
  message: 'yes',
  className: '',
  showMessage: jest.fn(),
  hideMessage: jest.fn(),
};

jest.useFakeTimers();

describe('Toast component', () => {
  it('should match component snapshot', () => {
    const tree = render.create(<Toast {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const shallowWrapper = shallow(<Toast {...props} />);
    expect(shallowWrapper).toBeDefined();
    expect(shallowWrapper.find('.toast__container').length).toBe(1);
    expect(shallowWrapper.exists()).toBe(true);
  });
});

describe('Toast Provider component', () => {
  it('should match component snapshot', () => {
    const tree = render.create(<ToastProvider {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const shallowWrapper = shallow(<ToastProvider {...props} />);
    expect(shallowWrapper).toBeDefined();
    expect(shallowWrapper.exists()).toBe(true);
  });

  it('should call showMessage method', () => {
    const shallowWrapper = shallow(<ToastProvider {...props} />);
    const showMessageSpy = spy(shallowWrapper.instance(), 'showMessage');
    expect(shallowWrapper.state().message).toEqual(null);

    shallowWrapper.instance().showMessage('hello', 'class');
    expect(showMessageSpy.called).toEqual(true);
    expect(shallowWrapper.state().message).toEqual('hello');
    expect(shallowWrapper.state().className).toEqual('class');
  });

  it('should call hideMessage method', () => {
    const shallowWrapper = shallow(<ToastProvider {...props} />);
    const hideMessageSpy = spy(shallowWrapper.instance(), 'hideMessage');
    shallowWrapper.setState({ message: 'hello' });

    shallowWrapper.instance().hideMessage();
    expect(shallowWrapper.state().message).toEqual('hello');
    expect(hideMessageSpy.called).toEqual(true);
    jest.runAllTimers();
    expect(shallowWrapper.state().message).toEqual(null);
  });
});
