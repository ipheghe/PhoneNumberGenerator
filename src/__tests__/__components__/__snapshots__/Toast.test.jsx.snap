// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast Provider component should match component snapshot 1`] = `null`;

exports[`Toast component should match component snapshot 1`] = `
<div
  className="toast__container "
>
  <p>
    yes
  </p>
</div>
`;
